tiledb/sm/stats/timer_stat.cc

```diff
@@ -0,0 +1,118 @@
+/**
+ * @file   timer_stat.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2018-2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file contains the TimerStat implementation.
+ */
+
+#include "tiledb/sm/stats/timer_stat.h"
+
+namespace tiledb {
+namespace sm {
+namespace stats {
+
+// Static member variable definitions.
+std::unordered_map<
+    std::thread::id,
+    std::chrono::high_resolution_clock::time_point>
+    TimerStat::start_times_;
+
+/* ****************************** */
+/*   CONSTRUCTORS & DESTRUCTORS   */
+/* ****************************** */
+
+TimerStat::TimerStat()
+    : duration_(0.0f) {
+}
+
+/* ****************************** */
+/*              API               */
+/* ****************************** */
+
+void TimerStat::start_timer(const std::thread::id tid) {
+  start_times_[tid] = std::chrono::high_resolution_clock::now();
```

Comment:
I believe #1994 addressed this, but to confirm, there is no way for 1 thread end up calling the same function that we have a timer on? Any recursion anywhere? Just want to make sure the thread id is unique and we won't run into problems assuming there can only be one timer per function per thread?

Reply:
This is a valid point!

Reply:
This is unrelated to recursion and unresolved by #1994. It is possible that two threads may both start the same timer. For instance, something like:
```
parallel_for(0, 2, []() {
  STATS_START_TIMER(stats::Stats::TimerType::FOO)
  sleep(10);
  STATS_END_TIMER(stats::Stats::TimerType::FOO)
});
```